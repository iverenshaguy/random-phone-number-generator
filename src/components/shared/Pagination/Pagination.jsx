import React, { Component } from 'react';
import ReactPagination from "react-js-pagination";
import PropTypes from 'prop-types';

import WindowsDimensionContext from '../../../contexts/WindowsDimensionContext';
import './Pagination.scss';


class Pagination extends Component {
  static contextType = WindowsDimensionContext;

  static propTypes = {
    onPageChange: PropTypes.func.isRequired,
    metadata: PropTypes.shape({
      page: PropTypes.number,
      perPage: PropTypes.number,
      totalCount: PropTypes.number,
    }),
  };

  state = {
    pages: [],
    currentPage: 1,
    lastPage: 1
  };

  static getDerivedStateFromProps(props) {
    const { metadata: {
      totalCount, page, perPage,
    } } = props;
    const numberOfPages = Math.ceil(totalCount/perPage);
    const pages = Array.from({ length: numberOfPages }, (v, i) => i + 1);

    return {
      pages,
      currentPage: page,
      lastPage: pages[pages.length - 1]
    };
  }

  setPage(page) {
    this.props.onPageChange(page);
  }

  render() {
    const { onPageChange, metadata: { perPage, page, totalCount } } = this.props;
    const { isMobileSized } = this.context;
    const { pages } = this.state;
    const classes = `mx-auto ${pages.length <= 1 ? 'invisible' : 'visible'}`;
    const pageRangeDisplayed = isMobileSized ? 5 : 10;

    return (
      <div className={classes}>
        <ReactPagination
          activePage={page}
          itemsCountPerPage={perPage}
          totalItemsCount={totalCount}
          pageRangeDisplayed={pageRangeDisplayed}
          onChange={onPageChange}
          itemClass="page-item"
        />
      </div>
    );
  }
}

export default Pagination;
