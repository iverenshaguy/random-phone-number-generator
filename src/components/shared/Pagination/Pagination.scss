.pagination {
  border: 1px solid #ced4da;
  border-radius: 0.25em;

  li {
    padding: 5px;
    min-width: 65px;
    border-right: 1px solid #ced4da;
    text-align: center;

    &:last-child {
      border-right: 0;
    }

    a {
      color: #495057;
    }
  }

  @media screen and (max-width: 992px) {
    li {
      padding: 5px;
      min-width: 35px;
    }
  }
}