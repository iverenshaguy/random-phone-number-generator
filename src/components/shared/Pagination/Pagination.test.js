import React from 'react';
import PropTypes from 'prop-types';

import Pagination from './Pagination';


describe('Pagination', () => {
  const props = {
    onPageChange: jest.fn(),
    metadata: {
      page: 1,
      perPage: 10,
      totalCount: 450,
    }
  };
  const wrapper = mount(
    <Pagination {...props} />, {
      context: {
        windowWidth: 1000,
        windowHeight: 1500,
        isMobileSized: false,
      },
      childContextTypes: {
        windowWidth: PropTypes.number,
        windowHeight: PropTypes.number,
        isMobileSized: PropTypes.bool,
      }
    }
  );

  it('should render Pagination component correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('calls setPage when a pagination item is clicked', () => {
    wrapper.find('a:first-child').at(7).simulate('click');
    expect(props.onPageChange).toHaveBeenCalled();
  });
});
