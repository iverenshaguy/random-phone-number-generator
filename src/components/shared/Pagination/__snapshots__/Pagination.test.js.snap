// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render Pagination component correctly 1`] = `
<Pagination
  metadata={
    Object {
      "page": 1,
      "perPage": 10,
      "totalCount": 450,
    }
  }
  onPageChange={[MockFunction]}
>
  <div
    className="mx-auto visible"
  >
    <a
      activePage={1}
      firstPageText="«"
      getPageUrl={[Function]}
      hideFirstLastPages={false}
      innerClass="pagination"
      itemClass="page-item"
      itemsCountPerPage={10}
      lastPageText="»"
      nextPageText="⟩"
      onChange={[MockFunction]}
      pageRangeDisplayed={10}
      prevPageText="⟨"
      totalItemsCount={450}
    >
      <ul
        className="pagination"
      >
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={true}
          itemClass="page-item"
          key="first"
          linkClass=""
          onClick={[MockFunction]}
          pageNumber={1}
          pageText="«"
        >
          <li
            className="page-item disabled"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              «
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={true}
          itemClass="page-item"
          key="prev0"
          linkClass=""
          onClick={[MockFunction]}
          pageNumber={0}
          pageText="⟨"
        >
          <li
            className="page-item disabled"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              ⟨
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={true}
          isDisabled={false}
          itemClass="page-item"
          key="1"
          onClick={[MockFunction]}
          pageNumber={1}
          pageText="1"
        >
          <li
            className="page-item active"
            onClick={[Function]}
          >
            <a
              className="undefined"
              href="#"
            >
              1
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="2"
          onClick={[MockFunction]}
          pageNumber={2}
          pageText="2"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              2
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="3"
          onClick={[MockFunction]}
          pageNumber={3}
          pageText="3"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              3
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="4"
          onClick={[MockFunction]}
          pageNumber={4}
          pageText="4"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              4
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="5"
          onClick={[MockFunction]}
          pageNumber={5}
          pageText="5"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              5
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="6"
          onClick={[MockFunction]}
          pageNumber={6}
          pageText="6"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              6
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="7"
          onClick={[MockFunction]}
          pageNumber={7}
          pageText="7"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              7
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="8"
          onClick={[MockFunction]}
          pageNumber={8}
          pageText="8"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              8
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="9"
          onClick={[MockFunction]}
          pageNumber={9}
          pageText="9"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              9
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="10"
          onClick={[MockFunction]}
          pageNumber={10}
          pageText="10"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              10
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="next2"
          linkClass=""
          onClick={[MockFunction]}
          pageNumber={2}
          pageText="⟩"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              ⟩
            </a>
          </li>
        </t>
        <t
          activeClass="active"
          disabledClass="disabled"
          href="#"
          isActive={false}
          isDisabled={false}
          itemClass="page-item"
          key="last"
          linkClass=""
          onClick={[MockFunction]}
          pageNumber={45}
          pageText="»"
        >
          <li
            className="page-item"
            onClick={[Function]}
          >
            <a
              className=""
              href="#"
            >
              »
            </a>
          </li>
        </t>
      </ul>
    </a>
  </div>
</Pagination>
`;
